### ISSUE TITLE
Reproducing Token and Sent error rates

### ISSUE THREAD

**Issue by tanay2001** [tagoyal/factuality-datasets#8]
Hi @tagoyal 

Awesome work! 
I am trying to reproduce the word and sent ER scores in Table 6 using this code base, specifically using `evaluate_generated_outputs.py`, Could you clarify how the Word ER and Sent ER relate to https://github.com/tagoyal/factuality-datasets/blob/c81fe64e02147d9ff7c50bff6f8e4965b0af189c/evaluate_generated_outputs.py#L82-L86
 and 
https://github.com/tagoyal/factuality-datasets/blob/c81fe64e02147d9ff7c50bff6f8e4965b0af189c/evaluate_generated_outputs.py#L91

Thanks!

**Comment by tagoyal** [comment]
hi, 
for sentence error rates, you can use the 'pred' values in the code above. I'll update the code to also compute Word ER by next week. 

Basically, you need to covert the arc level factuality labels to word level factuality: if all arcs origination from / to a word is non-factual, the word is considered to be non-factual. Otherwise factual. 

Thanks,
Tanya

**Comment by tanay2001** [comment]
Got it , Thanks!

**Comment by tagoyal** [comment]
hi tanay,

`def evaluate_summary(article_data, summary, tokenizer, model, nlp, args):
    eval_dataset = get_single_features(summary, article_data, tokenizer, nlp, args)
    eval_sampler = SequentialSampler(eval_dataset)
    eval_dataloader = DataLoader(eval_dataset, sampler=eval_sampler, batch_size=1)
    batch = [t for t in eval_dataloader][0]
    device = args.device
    batch = tuple(t.to(device) for t in batch)

    with torch.no_grad():
        input_ids, attention, child, head = batch[0], batch[1], batch[2], batch[3]
        mask_entail, mask_cont, num_dependency, arcs = batch[4], batch[5], batch[6], batch[7]
        sent_labels = batch[8]

        inputs = {'input_ids': input_ids, 'attention': attention,  'child': child,
                  'head': head, 'mask_entail': mask_entail, 'mask_cont': mask_cont, 'mask_neutral': mask_neutral,
                  'num_dependency': num_dependency, 'sent_label': sent_labels, 'device': device}

        outputs = model(**inputs)
        dep_outputs = outputs[1].detach()
        dep_outputs = dep_outputs.squeeze(0)
        dep_outputs = dep_outputs[:num_dependency, :].cpu().numpy()

        child = child[0, :num_dependency].cpu().numpy()
        head = head[0, :num_dependency].cpu().numpy()

        input_full = tokenizer.convert_ids_to_tokens(input_ids[0], skip_special_tokens=False)
        input_full = ' '.join(input_full).replace('[PAD]', '').strip()

        input = input_full.split('[SEP]')[0].strip().split(' ')
        summary = input_full.split('[SEP]')[1].strip().split(' ')
        len_input = len(input)

        dep_pair_dict = {}
        for k in range(num_dependency):
            pred = np.argmax(dep_outputs[k])
            child_idx = child[k] - len_input - 1
            head_idx = head[k] - len_input - 1

            if head_idx in dep_pair_dict.keys():
                dep_pair_dict[head_idx].append((child_idx, pred))
            else:
                dep_pair_dict[head_idx] = [(child_idx, pred)]

            if child_idx in dep_pair_dict.keys():
                dep_pair_dict[child_idx].append((head_idx, pred))
            else:
                dep_pair_dict[child_idx] = [(head_idx, pred)]

        wrong_ids = []
        for idx in dep_pair_dict:
            any_0 = False
            for dep in dep_pair_dict[idx]:
                if dep[1] == 0:
                    any_0 = True
                    break
            if any_0:
                wrong_ids.append(idx)

        id_actualid_mapping = {}
        actual_idx = 0
        for idx, word in enumerate(summary):
            if not summary[idx].startswith('##'):
                actual_idx += 1
            id_actualid_mapping[idx] = actual_idx - 1

        actual_wrong_ids = []
        for idx in wrong_ids:
            actual_wrong_ids.append(id_actualid_mapping[idx])

        actual_all_ids = []
        for idx in dep_pair_dict:
            actual_all_ids.append(id_actualid_mapping[idx])

        return actual_wrong_ids, actual_all_ids
`

**Comment by tagoyal** [comment]
`
Hi Tanay,
the code is shared below. you can replace the relevant code in evaluate_summary function with below. It returns the number of error words and total number of words in each summary. add these up for all summaries and that fraction is the final value reported.
BTW, I misremembered above, if a word has **any** non-factual arc, it was classified as non-factual. 



            def evaluate_summary(article_data, summary, tokenizer, model, nlp, args):
                  eval_dataset = get_single_features(summary, article_data, tokenizer, nlp, args)
                  eval_sampler = SequentialSampler(eval_dataset)
                  eval_dataloader = DataLoader(eval_dataset, sampler=eval_sampler, batch_size=1)
                  batch = [t for t in eval_dataloader][0]
                  device = args.device
                  batch = tuple(t.to(device) for t in batch)
                  with torch.no_grad():
                      input_ids, attention, child, head = batch[0], batch[1], batch[2], batch[3]
                      mask_entail, mask_cont, num_dependency, arcs = batch[4], batch[5], batch[6], batch[7]
                      sent_labels = batch[8]

                      inputs = {'input_ids': input_ids, 'attention': attention,  'child': child,
                                'head': head, 'mask_entail': mask_entail, 'mask_cont': mask_cont, 'mask_neutral': mask_neutral,
                                'num_dependency': num_dependency, 'sent_label': sent_labels, 'device': device}
              
                      outputs = model(**inputs)
                      dep_outputs = outputs[1].detach()
                      dep_outputs = dep_outputs.squeeze(0)
                      dep_outputs = dep_outputs[:num_dependency, :].cpu().numpy()
              
                      child = child[0, :num_dependency].cpu().numpy()
                      head = head[0, :num_dependency].cpu().numpy()
              
                      input_full = tokenizer.convert_ids_to_tokens(input_ids[0], skip_special_tokens=False)
                      input_full = ' '.join(input_full).replace('[PAD]', '').strip()
              
                      input = input_full.split('[SEP]')[0].strip().split(' ')
                      summary = input_full.split('[SEP]')[1].strip().split(' ')
                      len_input = len(input)
              
                      dep_pair_dict = {}
                      for k in range(num_dependency):
                          pred = np.argmax(dep_outputs[k])
                          child_idx = child[k] - len_input - 1
                          head_idx = head[k] - len_input - 1
              
                          if head_idx in dep_pair_dict.keys():
                              dep_pair_dict[head_idx].append((child_idx, pred))
                          else:
                              dep_pair_dict[head_idx] = [(child_idx, pred)]
              
                          if child_idx in dep_pair_dict.keys():
                              dep_pair_dict[child_idx].append((head_idx, pred))
                          else:
                              dep_pair_dict[child_idx] = [(head_idx, pred)]
              
                      wrong_ids = []
                      for idx in dep_pair_dict:
                          any_0 = False
                          for dep in dep_pair_dict[idx]:
                              if dep[1] == 0:
                                  any_0 = True
                                  break
                          if any_0:
                              wrong_ids.append(idx)
              
                      id_actualid_mapping = {}
                      actual_idx = 0
                      for idx, word in enumerate(summary):
                          if not summary[idx].startswith('##'):
                              actual_idx += 1
                          id_actualid_mapping[idx] = actual_idx - 1
              
                      actual_wrong_ids = []
                      for idx in wrong_ids:
                          actual_wrong_ids.append(id_actualid_mapping[idx])
              
                      actual_all_ids = []
                      for idx in dep_pair_dict:
                          actual_all_ids.append(id_actualid_mapping[idx])

                      return actual_wrong_ids, actual_all_ids`

**Comment by tanay2001** [comment]
Thanks a lot for updating the code!
